### PR TITLE
 [IMP] base, *: re-use env during registry loading

### PR DESCRIPTION
**Pull request by beledouxdenis**
The goal of this revision is to re-use the environment among the
different steps of the registry loading,
instead of creating a new environment for each step.

1. Simply To avoid to repeat the line
   `env = api.Environment(cr, SUPERUSER_ID, {})`
   multiple times in the code
2. This also allows to share the context among the different
   steps. This is not yet used in this revision, but it could
   be, for instance to avoid the current repetition to add the keys
   `install_module`, in `convert_csv_import` and `xml_import._tag_record`

### CODE REVIEW

**Review by Feyensv — APPROVED** [review]
LGTM
(a little note in the orm changelog of the documentation could be useful to explain the change ?)

**Comment by beledouxdenis** [comment]
@rco-odoo I would like your opinion regarding this change, especially regarding the fact to re-use the env during the different steps and accross modules during the installation in `loading.py`
https://github.com/odoo/odoo/pull/108254/files#diff-250c832b5654555cb37bc4e064c7d69e0c970ce2b0f5eba5301d94b9e973c8d2R417

I am not sure if there is/was a good technical reason to use a brand new environment between each module loading and each step, or if it's just historical and we didn't take the risk at the time to re-use the environment and just create a brand new one from `cr`.

**Comment by beledouxdenis** [comment]
@rco-odoo pretty please

File: odoo/modules/loading.py [hunk 1]
```diff
@@ -103,27 +102,26 @@ def load_demo(cr, package, idref, mode):
...
         "SELECT name FROM ir_module_module WHERE state IN ('installed', 'to upgrade', 'to remove')"
     )
-    module_list = [name for (name,) in cr.fetchall()]
-    graph.add_modules(cr, module_list, ['demo'])
+    module_list = [name for (name,) in env.cr.fetchall()]
+    graph.add_modules(env.cr, module_list, ['demo'])
 
     for package in graph:
-        load_demo(cr, package, {}, 'init')
+        load_demo(env.cr, package, {}, 'init')
```
Comment (rco-odoo) [inline odoo/modules/loading.py:118]:
```suggestion
        load_demo(env, package, {}, 'init')
```

**Review by rco-odoo — APPROVED** [review]
This "alignment" of those old APIs on using environments instead of simply a cursor totally makes sense, and should have been done years ago.  Besides the small bug I found below (calling `load_demo` with `env.cr` instead of `env`), things look good to me.

File: odoo/tools/convert.py [hunk 2]
```diff
@@ -647,14 +647,13 @@ def convert_csv_import(cr, module, fname, csvcontent, idref=None, mode='init',
         'install_filename': fname,
         'noupdate': noupdate,
     }
-    env = odoo.api.Environment(cr, SUPERUSER_ID, context)
-    result = env[model].load(fields, datas)
+    result = env[model].with_context(**context).load(fields, datas)
```
Comment (KangOl) [inline odoo/tools/convert.py:650]: It changes the behavior.
To keep the same as before:
```suggestion
    result = env[model].with_context(context).load(fields, datas)
```

Comment (beledouxdenis) [reply odoo/tools/convert.py:650]: Indeed, but is it a bad thing ?
For instance if you want to import a file with some flags in the context activated ? e.g `mail_create_nolog=True` ?

Comment (KangOl) [reply odoo/tools/convert.py:650]: We can keep it like this.

**Review by tde-banana-odoo — APPROVED** [review]

**Comment by KangOl** [comment]
@robodoo rebase-ff r+

**Comment by beledouxdenis** [comment]
robodoo override=ci/security

The getattr calls were already there, only the parameter to call the method returned by getattr is changed, from `cr` to `env`
```
odoo/odoo/modules/loading.py:191
getattr(py_module, pre_init)(env)
```

```diff
-                getattr(py_module, pre_init)(cr)
+                getattr(py_module, pre_init)(env)
```

```
odoo/odoo/modules/loading.py:244
getattr(py_module, post_init)(env)
```

```diff
-                    getattr(py_module, post_init)(cr, registry)
+                    getattr(py_module, post_init)(env)
```

```
odoo/odoo/modules/loading.py:533
getattr(py_module, uninstall_hook)(env)
```

```diff
-                        getattr(py_module, uninstall_hook)(cr, registry)
+                        getattr(py_module, uninstall_hook)(env)
```

```
odoo/odoo/tools/convert.py:620
env.cr.execute(fp.read()) # pylint: disable=sql-injection
```

The `cr.execute(fp.read())` was already there, `cr.execute` is changed to `env.cr.execute`, which is the same thing.

```diff
-def convert_sql_import(cr, fp):
-    cr.execute(fp.read()) # pylint: disable=sql-injection
+def convert_sql_import(env, fp):
+    env.cr.execute(fp.read()) # pylint: disable=sql-injection
```

**Comment by HydrionBurst** [comment]
I think this PR may have some problem about translated fields.
Since the `env` is passed to the `xml_import` the `env.lang` will be used to load the xml.
Previously, the `env.lang` is always `None` which makes sense and the loaded xml will be loaded as en_US
Now, it will become the translation for the user language

The question is should we ask developers to pass the correct `env` with_context `lang=en_US` or force the lang to en_US in xml_import

**Pull request closed without merging**